app/src/test/java/org/mozilla/tv/firefox/pocket/PocketVideoFetchWorkerTest.kt

```diff
@@ -0,0 +1,90 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.pocket
+
+import androidx.work.ListenableWorker.Result
+import androidx.work.WorkerParameters
+import io.mockk.MockKAnnotations
+import io.mockk.coEvery
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.mockk
+import io.mockk.verify
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.mozilla.tv.firefox.FirefoxApplication
+import org.mozilla.tv.firefox.utils.ServiceLocator
+
+class PocketVideoFetchWorkerTest {
+
+    private lateinit var worker: PocketVideoFetchWorker
+
+    @MockK private lateinit var context: FirefoxApplication
+    @Suppress("DEPRECATION") // We need endpoint raw until we move to the a-c impl.
+    @MockK private lateinit var endpointRaw: PocketEndpointRaw
+    @MockK private lateinit var store: PocketVideoStore
+
+    @MockK private lateinit var workerParams: WorkerParameters
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+
+        every { context.applicationContext } returns context
+        every { context.serviceLocator } returns mockk<ServiceLocator>().also {
+            every { it.pocketEndpointRaw } returns endpointRaw
+            every { it.pocketVideoStore } returns store
+        }
+
+        worker = PocketVideoFetchWorker(context, workerParams)
+    }
+
+    @Test
+    fun `WHEN the endpoint fails to return a result THEN the job will retry`() {
+        coEvery { endpointRaw.getGlobalVideoRecommendations() } returns null
+        assertEquals(Result.retry(), worker.doWork())
+    }
+
+    @Test
+    fun `WHEN the endpoint returns a non-null result THEN the store decides what happens to the exact json string`() {
+        arrayOf(
+            "",
+            "{",
+            "{}",
+            "{\"json\": \"yeah!\"}"
+        ).forEachIndexed { i, json ->
+            coEvery { endpointRaw.getGlobalVideoRecommendations() } returns json
+            every { store.save(any()) } returns true
+
+            worker.doWork()
+
+            println("For index $i: $json")
+            verify { store.save(json) }
+        }
+    }
+
+    @Test
+    fun `WHEN the store says the endpoint returned valid json THEN the job will fail`() {
+        // The store returns false if the JSON is invalid and refuses to be saved:
+        // unfortunately, this naming is not very intuitive.
+        everyEndpointRawGetVideoRecsReturnsANonNullValue()
+        every { store.save(any()) } returns false
+        assertEquals(Result.failure(), worker.doWork())
+    }
+
+    @Test
+    fun `WHEN the endpoint returns valid json THEN the job will succeed`() {
+        // The store returns true if the JSON is valid and can be saved:
+        // unfortunately, this naming is not very intuitive.
+        everyEndpointRawGetVideoRecsReturnsANonNullValue()
```

Comment:
I'm a little confused here. Doesn't this function return invalid json since it is empty?
I'm not fully understanding the difference between this test and the one above.

Reply:
The decision for invalid or valid JSON is made by the store, which returns true in this test whereas the last one returned false. The data returned by the endpoint doesn't seem to matter. I'm not sure if this test is well written or not. :\